Take the user through a bootstrap wizard when they close gitops run

opudrovs opened the pull request:
Closes #2699

WIP, needs code check and description.

opudrovs commented:
I will add UI screenshots or video now.

opudrovs commented:
UI screenshots. Different flag lists are used for different git providers. If remote URL contains `github.com` or `gitlab.com`, some values will be prefilled.

The select git provider table will be displayed only if the git provider vannot be parsed from the remote URL. It is run as a separate bubbletea program. Both programs exit to the main command flow when done. The UI uses default styling from bubbletea examples for now.

<img width="462" alt="Screenshot 2022-09-26 at 13 07 10" src="https://user-images.githubusercontent.com/8824708/192262473-9d392282-0a80-4d28-94bd-13bdf2cb17ee.png">
<img width="747" alt="Screenshot 2022-09-26 at 13 07 26" src="https://user-images.githubusercontent.com/8824708/192262478-ba5f14b7-5ceb-43b2-ae34-1ec1398b0ebc.png">
<img width="753" alt="Screenshot 2022-09-26 at 13 07 46" src="https://user-images.githubusercontent.com/8824708/192262481-c73b2d9f-b34a-40db-bfb1-7dceb760b906.png">
<img width="672" alt="Screenshot 2022-09-26 at 13 08 00" src="https://user-images.githubusercontent.com/8824708/192262485-9c403269-202f-47d1-8350-e7af4d527588.png">
<img width="762" alt="Screenshot 2022-09-26 at 13 08 16" src="https://user-images.githubusercontent.com/8824708/192262487-6322db1c-940c-424a-9b8b-e45caad2a101.png">
<img width="752" alt="Screenshot 2022-09-26 at 13 09 37" src="https://user-images.githubusercontent.com/8824708/192262490-98c321c9-bd4e-4364-a172-b451a27ff41a.png">
<img width="759" alt="Screenshot 2022-09-26 at 13 10 08" src="https://user-images.githubusercontent.com/8824708/192262491-fa8e8e82-b8dc-4b42-b872-fe4d804905ce.png">

opudrovs commented:
For testing the wizard UI easily I uncomment out previous gitops run code and uncomment checks around calls to`SelectGitProvider`.

chanwit commented:
It's so beautiful! Thank you so much @opudrovs

chanwit commented:
`path: ./clusters/<...>` is still missing from the generated command, AFAICT

opudrovs commented:
@chanwit 

> path: ./clusters/<...> is still missing from the generated command, AFAICT

Was the `path` flag value empty? If yes, it will be removed from flags for now, I don't handle required flags yet.

Or do you mean that it should it be autofilled from the gitops run commandline arguments?

ozamosi commented:
:tada: Beautiful!

:bulb: I would change `defaultFlagValue` from `string` to something like `type DefaultValueGetter func(*gogit.Repository) string` - that way, the auto-detection bits would sit together with their definitions, instead of the auto-detection being in `NewBootstrapWizard` separate from the definitions.

We could still use something like
```
func constantDefault(value string) DefaultValueGetter {
	return func(_ *gogit.Repository) string {
		return value
	}
}
```
so you only need to change `defaultFlagValue: "true"` to `defaultFlagValue: constantDefault("true")` - but we'd also be able to write more complex getters for e.g. path, branch, hit the github API to check if repo exists, use the OS username as default for owner, etc.

opudrovs commented:
Thanks! OK, will do it.

ozamosi commented on cmd/gitops/beta/run/cmd.go:
```diff
@@ -525,6 +527,72 @@ func betaRunCommandRunE(opts *config.Options) func(*cobra.Command, []string) err
 		}
 
+		// run bootstrap wizard only if Flux was not installed
+		if fluxVersion != "" {
```
:no_entry: Final thought: it's Release Week, and the follow up tickets won't be done before we want to ship. Could we just add a quick check here that, say, the `GITOPS_RUN_BOOTSTRAP` env var is set, and if not also return nil?

opudrovs replied:
Sure, not a problem.

Should I just check for `os.Getenv("GITOPS_RUN_BOOTSTRAP")` or do I also need a call to `osSetenv` somewhere?

ozamosi replied:
No setenv, just check.

opudrovs commented:
@ozamosi @chanwit 

 - Added check for os.Getenv("GITOPS_RUN_BOOTSTRAP") in addition to flux version before running the wizard.
- Replace default flag values with callbacks (except for path, did not create a getter for it because it's computed not from repo).
- Improve parsing of remote URLs.
- Added parsing branch as repo.Head().

opudrovs commented:
Mwahaha! I am merging it! 😈
